fastfield_codecs/src/frame_of_reference.rs

```diff
@@ -0,0 +1,303 @@
+}
+
+impl BlockReader {
+    fn new(metadata: BlockMetadata, start_offset: u64) -> Self {
+        Self {
+            bit_unpacker: BitUnpacker::new(metadata.num_bits),
+            metadata,
+            start_offset,
+        }
+    }
+
+    #[inline]
+    fn get_u64(&self, block_pos: u64, data: &[u8]) -> u64 {
+        let diff = self
+            .bit_unpacker
+            .get(block_pos, &data[self.start_offset as usize..]);
+        self.metadata.min + diff
+    }
+}
+
+impl BinarySerializable for BlockMetadata {
+    fn serialize<W: Write>(&self, write: &mut W) -> io::Result<()> {
+        self.min.serialize(write)?;
+        self.num_bits.serialize(write)?;
+        Ok(())
+    }
+
+    fn deserialize<R: Read>(reader: &mut R) -> io::Result<Self> {
+        let min = u64::deserialize(reader)?;
+        let num_bits = u8::deserialize(reader)?;
+        Ok(Self { min, num_bits })
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct FrameOfReferenceFooter {
+    pub num_vals: u64,
+    pub min_value: u64,
+    pub max_value: u64,
+    block_metadatas: Vec<BlockMetadata>,
+}
+
+impl BinarySerializable for FrameOfReferenceFooter {
+    fn serialize<W: Write>(&self, write: &mut W) -> io::Result<()> {
+        let mut out = vec![];
+        self.num_vals.serialize(&mut out)?;
+        self.min_value.serialize(&mut out)?;
+        self.max_value.serialize(&mut out)?;
+        self.block_metadatas.serialize(&mut out)?;
+        write.write_all(&out)?;
+        (out.len() as u32).serialize(write)?;
+        Ok(())
+    }
+
+    fn deserialize<R: Read>(reader: &mut R) -> io::Result<Self> {
+        let footer = Self {
+            num_vals: u64::deserialize(reader)?,
+            min_value: u64::deserialize(reader)?,
+            max_value: u64::deserialize(reader)?,
+            block_metadatas: Vec::<BlockMetadata>::deserialize(reader)?,
+        };
+        Ok(footer)
+    }
+}
+
+impl FastFieldCodecReader for FrameOfReferenceFastFieldReader {
+    /// Opens a fast field given a file.
+    fn open_from_bytes(bytes: &[u8]) -> io::Result<Self> {
+        let footer_len: u32 = (&bytes[bytes.len() - 4..]).deserialize()?;
+        let (_, mut footer) = bytes.split_at(bytes.len() - (4 + footer_len) as usize);
+        let footer = FrameOfReferenceFooter::deserialize(&mut footer)?;
+        let mut block_readers = Vec::with_capacity(footer.block_metadatas.len());
+        let mut current_data_offset = 0;
+        for block_metadata in footer.block_metadatas.into_iter() {
+            let num_bits = block_metadata.num_bits;
+            block_readers.push(BlockReader::new(block_metadata, current_data_offset));
+            current_data_offset += num_bits as u64 * BLOCK_SIZE / 8;
+        }
+        Ok(Self {
+            num_vals: footer.num_vals,
+            min_value: footer.min_value,
+            max_value: footer.max_value,
+            block_readers,
+        })
+    }
+
+    #[inline]
+    fn get_u64(&self, doc: u64, data: &[u8]) -> u64 {
```

Comment:
doc -> DocId (which is a u32 actually)

Reply:
This is defined at the trait level, I prefer to change that in another PR.

Reply:
Actually scratch that!

This is not necessarily a doc here, which is why this is u64.
The multivalued fast field rely on 2 fast field as you know.

In this case the second one is *not* a DocId. Let's call doc "idx" or something in this function.
We should fix other fastfields as well as the trait.

Reply:
Ok I totally forgot about multivalued ff, I updated the trait.